hir2/src/patterns/applicator.rs

```diff
@@ -0,0 +1,179 @@
+
+/// This type manages the application of a group of rewrite patterns, with a user-provided cost model
+pub struct PatternApplicator {
+    /// The list that owns the patterns used within this applicator
+    rewrite_patterns_set: Rc<FrozenRewritePatternSet>,
+    /// The set of patterns to match for each operation, stable sorted by benefit.
+    patterns: BTreeMap<OperationName, SmallVec<[Rc<dyn RewritePattern>; 2]>>,
+    /// The set of patterns that may match against any operation type, stable sorted by benefit.
+    match_any_patterns: SmallVec<[Rc<dyn RewritePattern>; 1]>,
+}
+impl PatternApplicator {
+    pub fn new(rewrite_patterns_set: Rc<FrozenRewritePatternSet>) -> Self {
+        Self {
+            rewrite_patterns_set,
+            patterns: Default::default(),
+            match_any_patterns: Default::default(),
+        }
+    }
+
+    /// Apply a cost model to the patterns within this applicator.
+    pub fn apply_cost_model<CostModel>(&mut self, model: CostModel)
+    where
+        CostModel: Fn(&dyn RewritePattern) -> PatternBenefit,
+    {
+        // Clear the results computed by the previous cost model
+        self.match_any_patterns.clear();
+        self.patterns.clear();
+
+        // Filter out op-specific patterns with no benefit, and order by highest benefit first
+        let mut benefits = Vec::default();
+        for (op, op_patterns) in self.rewrite_patterns_set.op_specific_patterns().iter() {
+            benefits
+                .extend(op_patterns.iter().filter_map(|p| filter_map_pattern_benefit(p, &model)));
+            benefits.sort_by_key(|(_, benefit)| *benefit);
+            self.patterns
+                .insert(op.clone(), benefits.drain(..).map(|(pat, _)| pat).collect());
+        }
+
+        // Filter out "match any" patterns with no benefit, and order by highest benefit first
+        benefits.extend(
+            self.rewrite_patterns_set
+                .any_op_patterns()
+                .iter()
+                .filter_map(|p| filter_map_pattern_benefit(p, &model)),
+        );
+        benefits.sort_by_key(|(_, benefit)| *benefit);
+        self.match_any_patterns.extend(benefits.into_iter().map(|(pat, _)| pat));
+    }
+
+    /// Apply the default cost model that solely uses the pattern's static benefit
+    #[inline]
+    pub fn apply_default_cost_model(&mut self) {
+        self.apply_cost_model(|pattern| pattern.benefit());
+    }
+
+    /// Walk all of the patterns within the applicator.
+    pub fn walk_all_patterns<F>(&self, mut callback: F)
+    where
+        F: FnMut(Rc<dyn RewritePattern>),
+    {
+        for patterns in self.rewrite_patterns_set.op_specific_patterns().values() {
+            for pattern in patterns {
+                callback(Rc::clone(pattern));
+            }
+        }
+        for pattern in self.rewrite_patterns_set.any_op_patterns() {
+            callback(Rc::clone(pattern));
+        }
+    }
+
+    pub fn match_and_rewrite<A, F, S>(
```

Comment:
This looks astonishing! Looking forward to `RewritePattern` impl example.

Reply:
I've gotten virtually of the rewrite infra built now, but the last kink I'm working out is how to approach the way the `Listener` classes are used in MLIR. In effect, the way it is used in MLIR relies on mutable aliases, which is obviously a no go in Rust, and because of that (and I suspect somewhat as an artifact of the design they chose), the class hierarchy and interactions between builders/rewriters/listeners is _very_ confusing.

I've untangled the nest a bit in terms of how it actually works in MLIR, but its a non-starter in Rust, so a different design is required, but one that ideally achieves similar goals. The key thing listeners enable in MLIR, is the ability for say, the pattern rewrite driver, to be notified whenever a rewrite modifies the IR in some way, and as a result, modify its own state to take those modifications into account. There are other things as well, such as listeners which perform expensive debugging checks without those checks needing to be implemented by each builder/rewriter, but the main thing I care about is the rewrite driver.

There are a couple of approaches that I think should work, but need to test them out first. In any case, that's the main reason things have been a bit slow in the last week.